### DISCUSSION TITLE
Teams AI library(JS/C#) - General Availability!

### DISCUSSION THREAD

**Discussion by corinagum** [microsoft/teams-sdk#1060] [Announcements]
# Versions 1.0.0 released

We are thrilled to announce that the Teams AI Library (JS/C#), which was initially introduced in public preview on May 5th, has now reached General Availability! This marks a significant milestone in our journey to empower developers with a powerful platform for building intuitive, AI-driven applications within the Teams ecosystem. The updated SDK brings a host of enhancements designed to simplify the process of developing conversational apps, making it more developer-friendly and feature-rich than ever before. 

![Teams AI screencap](https://github.com/microsoft/teams-ai/assets/14900841/1238cb3f-e2aa-4133-94ff-a3ebc0089196)

In response to valuable user feedback and continuous improvements, we have streamlined several aspects of the SDK. 

- Simplified state management by using `TurnState` object, eliminating the need for a `TurnStateManager` class. 
- All prompt-related tasks are now handled by the `Planner`, offering better token management and reducing the likelihood of overflowing the model's context window.
- Introducing **augmentations**, which allows developers to add named augmentations to their prompts, making prompt engineering tasks more convenient.  
- Build FAQ & QnA AI apps via DataSource plugin type simplifies the integration of RAG (Retrieval-Augmented Generation) into prompts, enhancing their capabilities.
- Integration of AlphaWave, which gives the AI library the ability to repair malformed responses from the model. JSON Schema-based Validator plugins are now used to validate all structured responses, ensuring consistency and reliability
- The new `ActionPlanner` unifies all these features into a single planner that developers can pass to the Application object, streamlining the development process. 
 
## Resources:
 
- [JS GitHub release](https://github.com/microsoft/teams-ai/releases/tag/js-1.0.1) | [npmjs package](https://www.npmjs.com/package/@microsoft/teams-ai) 
- [C# release](https://github.com/microsoft/teams-ai/releases/tag/dotnet-1.0.0) | [Nuget package](https://www.nuget.org/packages/Microsoft.Teams.AI)
- [Get started with Teams AI library - Microsoft Learn Documentation](https://learn.microsoft.com/microsoftteams/platform/bots/how-to/teams%20conversational%20ai/how-conversation-ai-get-started)

**Comment by Stevenic** [comment]
Great job team!!!!

**Reply by corinagum** [reply]
We miss you Steve! Looking forward to having you back! haha.

**Comment by mingjun1120** [comment]
Will there be a **Python** library in the future, in addition to the existing JavaScript and C# libraries?

**Reply by corinagum** [reply]
Yes! Python is currently in development, and we hope to announce public preview within the next couple of weeks. Thanks in advance for your patience!

**Comment by donatas-xyz** [comment]
Hi and thank you for a great release! 

[I'm struggling to find](https://stackoverflow.com/questions/77808829/how-to-chain-prompts-in-teams-ai-version-1-0-0-ms-teams-ai-bot-app) how are the chained prompts supposed to work? The `AI.chain()` method has now disappeared and I can't find information on what should I replace it with in version 1.0.0? Thank you! 

```typescript 
/**
     * Chains into another prompt and executes the plan that is returned.
     * @summary
     * This method is used to chain into another prompt. It will call the prompt manager to
     * get the plan for the prompt and then execute the plan. The return value indicates whether
     * that plan was completely executed or not, and can be used to make decisions about whether the
     * outer plan should continue executing.
     * @param {TurnContext} context Current turn context.
     * @param {TState} state Current turn state.
     * @param {string | PromptTemplate} prompt Optional. Prompt name or prompt template to use. If omitted, the AI systems default prompt will be used.
     * @param {Partial<AIOptions<TState>>} options Optional. Override options for the prompt. If omitted, the AI systems configured options will be used.
     * @returns {boolean} True if the plan was completely executed, otherwise false.
     */
    public async chain(
        context: TurnContext,
        state: TState,
        prompt?: string | PromptTemplate,
        options?: Partial<AIOptions<TState>>
    ): Promise<boolean> {}
```

**Reply by corinagum** [reply]
hi @BlueWhaleSEO, yes, `chain()` has been deprecated. Please take a look at the [chained actions](https://github.com/microsoft/teams-ai/blob/0fca2ed09d327ecdc682f2b15eb342a552733f5e/js/samples/04.ai.d.chainedActions.listBot/src/index.ts#L153) sample for a look on how to handle chain-of-thought. 

Please take a look at Monologue Augmentation. We also have a [PR](https://github.com/microsoft/teams-ai/pull/1176) that is currently in review that will introduce a lot of concepts, including how to handle chaining now that `chain` is deprecated.

**Reply by donatas-xyz** [reply]
Thank you @corinagum. I will have a look at this, although I've already searched extensively through the newest examples and the code changes past `.chain()`'s removal. I have no issues executing additional actions, but I guess the main part missing in my understanding is how do I execute those actions under different prompt instructions and different config (temperatures, token sizes etc)? I could achieve it nicely with `.chain` - i.e. a general chat prompt with it's own parameters then would lead to document generation prompt with its unique parameters too. Now everything seems to stay under the same prompt. 

The only similar [example ](https://github.com/microsoft/teams-ai/blob/main/js/samples/04.ai.b.messageExtensions.AI-ME/src/index.ts) to the actual prompt "chaining" I found is using `await planner.completePrompt(context, state, prompt);`, which I've tried, but all I'm getting back is DO/SAY response, which I believe may be caused by the #1130 bug. 

Because of this, I cannot progress with version 1.0.0 and had to go back to the original preview version, which still works. 

Hopefully I will get my light bulb moment soon enough.

**Reply by corinagum** [reply]
@BlueWhaleSEO I'm sorry to hear you've been having trouble. Why don't we do a couple of things:

1. Could you file a new Discussion item under the Questions category regarding this topic? This will help us track helping you out.
2. In that new discussion, if you could add code snippets and details of what you are working on, so I can have a detailed understanding of your goals, that would be much appreciated. If you can get me view access to your code, even better :)
3. I have a high priority work item that I need to work on today, so in the meantime, I recommend reading through the concept docs to have a better understanding of our updates in version 
4. Take a look at [Prompt templates](https://github.com/microsoft/teams-ai/blob/main/getting-started/CONCEPTS/PROMPTS.md#creating-prompt-templates) for multiple prompts and setting up `config.json`
5. Consider function calls `{{functionName}}` for sequencing prompts?

**Reply by donatas-xyz** [reply]
As suggested, I've moved the discussion here: https://github.com/microsoft/teams-ai/discussions/1322, @corinagum 
Thank you!
